[RDY] iproute: copy files in /etc

teto opened the pull request:
When doing source routing/multihoming, it's practical to give names to routing
tables. The absence of the rt_table file in /etc make this impossible.
This patch recreates these files on rebuild so that they can be modified
by the user

###### Motivation for this change
 see NixOS#38638.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39536&attempt_id=2ad772d5-acb9-4da1-9993-750ed5d3454a)

Attempted: iproute

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/wk7fql3ag8aq1n2zy5v7r25h2cjnfqr0-iproute2-4.16.0
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39536&attempt_id=c3c66eed-7615-434b-8929-0d4cd8f18f0a)

Attempted: iproute

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/07jzv5nwz5p1rn7nmw9fnf9gikir7ivg-iproute2-4.16.0
```
</p></details>

Mic92 commented on nixos/modules/system/activation/activation-script.nix:
```diff
@@ -119,6 +119,12 @@ in
     system.activationScripts.stdio = ""; # obsolete
 
+    system.activationScripts.iproute2 = ''
+      cp -R ${pkgs.iproute}/etc/iproute2 /etc
```
but does this not overwrite any modification done by the user?

andir replied:
I that case, would it not be better to have config options for those files? This feels wrong. The files should either be managed by nix or not touched like that at all..

Mic92 replied:
I agree with that.

teto replied:
once iproute is installed, users expect `ip route list table main` to work but if iproute can't find the file than it breaks this expection. While the file can be modified, it does have default. Maybe iproute should be patched look somewhere else than /etc/iproute2 (like in /var/run) but some sort of default should exist. To prevent overwriting user changes, the activationScript could write it only if it does'nt already exist.

teto commented:
yes it is a problem but i couldn't think of any better solution. in my case
restarting networkmanager should be enough to fix the rt_table file/the
routing tables (i update them through hm hooks).

Le ven. 27 avr. 2018 19:09, Jörg Thalheim <notifications@github.com> a
écrit :

> *@Mic92* commented on this pull request.
> ------------------------------
>
> In nixos/modules/system/activation/activation-script.nix
> <https://github.com/NixOS/nixpkgs/pull/39536#discussion_r184639227>:
>
> > @@ -119,6 +119,12 @@ in
>
>      system.activationScripts.stdio = ""; # obsolete
>
> +    system.activationScripts.iproute2 = ''
> +      cp -R ${pkgs.iproute}/etc/iproute2 /etc
>
> but does this not overwrite any modification done by the user?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/39536#pullrequestreview-115877210>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA2FOrAZh1vfhjhfsRkT5GPMGfWNDNkjks5tsudPgaJpZM4Tk6yt>
> .
>

matthewbauer commented:
Hm... Does iproute follow any environment variables to change the location of these things?

teto commented:
nope but you can configure an alternate path to look for on compilation (like /var/run).

matthewbauer commented:
Maybe we should patch it to look in ```/run/current-system/sw/etc```. That would be a little more consistent with NixOS.

To have /etc/iproute2 installed in /run/current-system/sw, you will need to add it to patchsToLink here:

https://github.com/NixOS/nixpkgs/blob/b6ec6fd52531a3f3f4dd0ee63df758beadd91cb5/nixos/modules/config/system-path.nix#L103-L125

teto commented:
For my specific scenario, the file is modified dynamically depending on network conditions. If a network pops up, it will create its own routing table alias in /etc/iproute2/rt_tables. So the file needs to be writable (ideally reset on boot), I don't think I can do this with pathsToLink.

matthewbauer commented:
Ok. This would probably make sense as a module then. It definitely should not be done by default. Or at least not until other users start complaining about the missing routing tables.

teto committed:
iproute2: module to create rt_table file & co

When doing source routing/multihoming, it's practical to give names to routing
tables. The absence of the rt_table file in /etc make this impossible.
This patch recreates these files on rebuild so that they can be modified
by the user see NixOS#38638.

iproute2 is modified to look into config.networking.iproute2.confDir instead of
/etc/iproute2.

teto commented:
I made it a module, I can merge it with the networking module if need be.

matthewbauer commented on nixos/modules/config/iproute2.nix:
```diff
@@ -0,0 +1,20 @@
+
+let
+  cfg = config.networking.iproute2;
+  runDir = "/run/iproute2";
```
Not to nitpick too much, but why not make this a config option like ```nixpkgs.config.iproute2.confdir``` & use it in place of runDir.

matthewbauer commented on pkgs/os-specific/linux/iproute/default.nix:
```diff
@@ -26,8 +26,9 @@ stdenv.mkDerivation rec {
+  # enable iproute2 module if you want this folder to be created
   buildFlags = [
-    "CONFDIR=/etc/iproute2"
+    "CONFDIR=/run/iproute2"
```
You can share this with the module by using something like ```config.iproute2.confdir or "/etc/iproute2"```

teto replied:
I was looking for a way to unify the setting too but not sure how to do it. I just tried sthg but it fails. Any suggestion ? Thanks for your patience

matthewbauer replied:
Yes I think you would do this like:

```
"CONFDIR=${config.iproute2.confdir or "/run/iproute2"}"
```

& then add config as an arg.

Also add to your module implementation:

```
nixpkgs.config.iproute2.confdir = "/run/iproute2"
```

teto replied:
just making sure:
- does it have to be a toplevel config.iproute2 or is the current config.networking.iproute2 ok ? 
- I already have confDir as an option ? isn't that enough ?

teto commented:
@matthewbauer is that conform to what you requested, if yes feel free to squash & merge.
many packets depend on iproute so I am interested in having this small change upstream.

matthewbauer commented on pkgs/os-specific/linux/iproute/default.nix:
```diff
@@ -26,8 +26,9 @@ stdenv.mkDerivation rec {
+  # enable iproute2 module if you want this folder to be created
   buildFlags = [
-    "CONFDIR=/etc/iproute2"
+    "CONFDIR=${config.networking.iproute2.confDir or "/run/iproute2"}"
```
Can you verify if this works? That is can you change confDir and get a different value passed to buildFlags? I'm not 100% on this but I think you need to directly pass it with ```nixpkgs.config``` in your module (see https://github.com/NixOS/nixpkgs/blob/release-18.03/nixos/modules/misc/nixpkgs.nix#L92-L106). Nixpkgs & NixOS have different config sets which can be confusing.

andir replied:
IMHO this requires a change log entry.

I am also not so sure if changing the default to `/run` is a good decision. If someone needs this the configuration will be set to whatever the user wants and the default behavior should not be altered if not used.

Am I missing something?

teto replied:
I agree this can be cool to have as a changelogentry. How would I do that ? other PR ? what document ?
I don't see a usecase either to let the user set the path (he would lose the cache) so I am tempted to remove the `confDir = mkOption {` but the paths between the module and the package must be in sync so I would let the nixpkgs.config.iproute2.confDir. If the user can't choose confDir, then we need to agree on the default. /etc/iproute2 is the common one but I guess /run/iproute2 matches dynamic files better. I am fine either way.

matthewbauer replied:
Ok yeah just using /run/iproute2 makes sense.

matthewbauer replied:
It's fine now

volth replied:
Why not to keep "/etc/iproute2" as the default in the case if ```config.networking.iproute2.confDir``` ?

I have a nixos module which creates files in "/etc/iproute2"

volth replied:
> /etc/iproute2 is the common one but I guess /run/iproute2 matches dynamic files better. I am fine either way.

Besides the compatibility with older Nixos and with the scripts made for other Linuxes, "/etc/iproute2" has the advantage of being configured via
```
environment.etc."iproute2".text = ''blablabla'';
```
while modification of "/run/iproute2" would require a ```systemd``` module with a carefully set up start order.

teto commented:
I tested it and the changes took effect. I squashed all.

teto committed:
release notes: mention iproute2 module

volth commented:
why not just 
```
environment.etc."iproute2/rt_tables" = {
  text = lib.fileContents "${pkgs.iproute}/etc/iproute2/rt_tables";
  mode = "0644";
};
```
instead of introducing so many configurable options ?

teto commented:
That sounds better, I don't think we need to have a configurable path. But I think this should be done by nixos so the module may still make sense. While I am just interested in rt_tables the other files in etc/iproute2/ should be copied too.
Is there a way to do this
```
environment.etc."iproute2/rt_tables" = {
  text = (lib.fileContents "${pkgs.iproute}/etc/iproute2/rt_tables") + '' any extras '';
  mode = "0644";
};
```
but for the whole `${pkgs.iproute}/etc/iproute2` folder ?

andir commented:
You could iterate through that dir using builtins.readDir
"${pks.iproute}/etc/iproute2/". That will give you an attrset where the
key/name is the filename and the value is the type of the dir.


On Fri, 8 Jun 2018, 05:37 Matthieu Coudron, <notifications@github.com>
wrote:

> That sounds better, I don't think we need to have a configurable path. But
> I think this should be done by nixos so the module may still make sense.
> While I am just interested in rt_tables the other files in etc/iproute2/
> should be copied too.
> Is there a way to do this
>
> environment.etc."iproute2/rt_tables" = {
>   text = (lib.fileContents "${pkgs.iproute}/etc/iproute2/rt_tables") + '' any extras '';
>   mode = "0644";
> };
>
> but for the whole ${pkgs.iproute}/etc/iproute2 folder ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/39536#issuecomment-395636317>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAm_dCncuDhWMizQbIZG8VfK-X940oNLks5t6fFlgaJpZM4Tk6yt>
> .
>

teto commented:
doesn't seem simpler than the current `cp -R`. 
To sum up the requested change would be to default back to /etc/iproute2 (and even remove the configurability).  @volth feel free to open a PR, else I can do it.

volth commented:
@teto 
I was thinking for doing the copy as part of the module which requires it (that one which sets up multi-homing routing).

When there will be 2+ modules which needs files in "/etc/iproute2" and they can be used simultaneously, then there will be sense for such module as "nixos/modules/config/iproute2.nix" with join'able ```config.iptables.rttablesExtraConfig``` et cetera (much like the other files in /etc are constructed by joining settings from many modules).

So, unless you want to contribute the multi-homing module, I am just for undoing PR https://github.com/NixOS/nixpkgs/pull/39536

teto commented:
What is that multihoming module ? I see your point in having  the file generated.

As for myself I use networkmanage hooks to achieve dynamic multihoming  https://github.com/NixOS/nixpkgs/pull/39142 so I would need `mode = "0644"`;
